active/0051-strategies.md

```diff
@@ -0,0 +1,247 @@
++ **File System Resolver** - Responsible for connecting a module name with a file path.
+
+`Linker` would be responsible for:
+
++ building a minimal dependency graph as well as check for redundant edges in the graph (more on the topic, [Transitive reduction of a directed graph](https://en.wikipedia.org/wiki/Transitive_reduction#Graph_algorithms_for_transitive_reduction));
++ producing an application tree with only used modules
+
+Dependency graph represents dependencies using module names, there is a need to be able to convert module name to file path. This is where `File System Resolver` comes in. Here's couple of examples:
+
+```javascript
+fileSystemResolver.resolve('lodash') => `some-path/node_modules/lodash/lodash.js`
+fileSystemResolver.resolve('ember-ajax') => `some-path/addon-modules/ember-ajax/index.js`
+fileSystemResolver.resolve('ember-data') => `some-path/addon-modules/modules/ember-data/index.js`
+fileSystemResolver.resolve('ember-data/-private') => `some-path/addon-modules/modules/-private.js`
+```
+
+This effort could be broken down into several phases:
+
++ dead modules elimination inside of the `addons/` (application would be the main entry point and unused modules are removed only from `addons/`)
++ dead modules elimination inside of the `app/`
+  + removing unused components and helpers (requires analysing templates)
+  + removing unused initializers/services (this likely entails work on dependency injection layer as we would need access to a resolver resolution map)
++ treeshaking (Rollup-like treeshaking where we include *only* the code that is used)
+
+`Linker` would be able to take an `exclude` list of modules as a parameter. Although, valuable in some situations, it should be clearly marked as advanced API. It should be used as a last resort and serve as an "escape hatch".
+
+It would make sense to implement `Linker` as a strategy. Developers would be able to "opt in"/"opt out" of optimising behaviour.
+
+## Deprecating `app.import` API
+
+Ember applications which choose to use `Linker` strategy should be able to remove usages of `app.import`.
+
+## Tools
+
+With growing complexity of Ember applications, it is crucial to provide more insights into final assets.
+
+Main goals are:
+
+- report raw/uglified/compressed asset sizes; [broccoli-concat-analyser](https://github.com/stefanpenner/broccoli-concat-analyser)
+- find source code duplication across your javascript assets (enables you to
+  fine tune code splitting parameters to reduce bundle invalidation rates as well
+  as improve repeat page load performance)
+
+# How We Teach This
+
+This is a backward compatible change to the existing Ember CLI eco system. In order to teach users how to use strategies, we need to update the API docs with a section for this and the best practices of when to use this. A more general purpose blog post could be beneficial as well.
+
+# Drawbacks
+
+- More libraries to maintain
+- Increasing public API surface
```

Comment:
I encourage you to think through this section more instead of only providing it lip service. Some ideas to start with:

* Does this feature further couple Ember apps and addons to Broccoli? Is that acceptable or a risk we can mitigate?
* `defaultStrategies` is an array in your example above. Is it ok that someone could `push` onto that array? Should we consider freezing the object or instead exporting a function that generates a new array on each invocation?
* Are there existing addons dependent on the current bundling implementation in Ember-CLI?
* Are there build performance concerns?
* Are the files being bundled only JavaScript, for example post HBS compilation? How can we bundle intelligently when we have a mix of HBS and JS files, or is that concern handled elsewhere?

I see you have some of these under unresolved, maybe some eventually will become drawbacks.

Reply:
all valid points and @rwjblue and I discussed all of them. having said that, we definitely need to have them in writing. let me address that.